sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NonAsyncTaskShouldNotReturnNull.cs

```diff
@@ -0,0 +1,114 @@
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        private static ISet<KnownType> TaskTypes =
+            new HashSet<KnownType>
+            {
+                KnownType.System_Threading_Tasks_Task,
+                KnownType.System_Threading_Tasks_Task_T
+            };
+
+        private static ISet<SyntaxKind> TrackedNullLiteralLocations =
+            new HashSet<SyntaxKind>
+            {
+                SyntaxKind.ArrowExpressionClause,
+                SyntaxKind.ReturnStatement,
+            };
+
+        private static ISet<SyntaxKind> PossibleEnclosingSyntax =
+            new HashSet<SyntaxKind>
+            {
+                SyntaxKind.VariableDeclaration,
+                SyntaxKind.PropertyDeclaration,
+                SyntaxKind.MethodDeclaration
+            };
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    var nullLiteral = (LiteralExpressionSyntax)c.Node;
+
+                    if (!nullLiteral.Parent.IsAnyKind(TrackedNullLiteralLocations))
+                    {
+                        return;
+                    }
+
+                    var enclosingMember = nullLiteral.Ancestors().FirstOrDefault(n => n.IsAnyKind(PossibleEnclosingSyntax));
+                    if (enclosingMember == null ||
+                        enclosingMember.IsKind(SyntaxKind.VariableDeclaration))
+                    {
+                        return;
+                    }
+
+                    var enclosingMemberSymbol = c.SemanticModel.GetDeclaredSymbol(enclosingMember);
+
+                    if (enclosingMemberSymbol == null ||
+                        !IsTaskReturnType(enclosingMemberSymbol) ||
+                        IsSafeTaskReturnType(enclosingMemberSymbol))
+                    {
+                        return;
+                    }
+
+                    c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, nullLiteral.GetLocation()));
+                },
+                SyntaxKind.NullLiteralExpression);
+        }
+
+        private static bool IsTaskReturnType(ISymbol symbol) =>
+            (GetReturnType(symbol) as INamedTypeSymbol)?.ConstructedFrom.DerivesFromAny(TaskTypes)
+            ?? false;
+
+        private static ITypeSymbol GetReturnType(ISymbol symbol) =>
+            symbol is IMethodSymbol methodSymbol
+            ? methodSymbol.ReturnType
+            : symbol.GetSymbolType();
+
+        private bool IsSafeTaskReturnType(ISymbol enclosingMemberSymbol) =>
+            // IMethodSymbol also handles lambdas
+            (enclosingMemberSymbol as IMethodSymbol)?.IsAsync
+            ?? false;
```

Comment:
(cosmetic) I would replace with `enclosingMemberSymbol is IMethodSymbol methodSymbol && methodSymbol.IsAsync`

This comment also applies for the `IsTaskReturnType` method. In general I don't like `?? false` constructs when they contain `as` operator and conditional member access. To me they seem less readable than a straight `null` check.